### ISSUE TITLE
Simpler version for Android?

### ISSUE THREAD

**Issue by brodybits** [sqlcipher/sqlcipher#125]
The existing Android version was a truly heroic undertaking and I really like using it. However, I find its API to be more complex than really necessary for many sqlite users and also becoming a challenge everyone to maintain. I found a couple alternatives that _could_ be easier for people to document and maintain a sqlcipher version of:
- https://code.google.com/p/sqlite3-android/
- https://code.google.com/p/sqlite4java/

In addition, I made an experimental, undocumented project that provides an extremely simple interface between Java and sqlite3 C library using Gluegen here: https://github.com/sqlg/SQLiteGlue

NOTE: This issue should be mutually exclusive to sqlcipher/android-database-sqlcipher#178 that I made to update the existing Android version.

**Comment by developernotes** [comment]
@brodybits - interesting, I had not heard of Gluegen before.  Something like this could certainly be beneficial for creating a lite library.  Have you used your version on any projects yet?

**Comment by brodybits** [comment]
Adding a couple more Java sqlite libraries, which seem to provide JDBC functionality, before I lose the links:
- https://github.com/crawshaw/sqlitejdbc - looks very old, some updates in https://github.com/evgeny-myasishchev/sqlitejdbc/tree/3.7.10, looks like MIT or ISC license
- http://www.ch-werner.de/javasqlite/ - looks up-to-date, looks like BSD-style license

**Comment by brodybits** [comment]
About https://github.com/sqlg/SQLiteGlue, which I published with Unlicense (public domain, based off sqlite license statement), I did try integrating it with my Cordova sqlite plugin a couple years ago, on an experimental basis. Unfortunately https://github.com/sqlg/SQLiteGlue does (like any other Java/Android sqlite library) make a hard dependency on a certain implementation of sqlite, which could be changed at compile-time or link-time but NOT runtime.

Yesterday I made a set of library classes in https://github.com/brodybits/sqlc-test, also public domain, that define the following interfaces and classes:
- `SQLiteConnection` - Java interface with basic operation functions on a database connection object such as `close()`, `getTotalChanges()`, and `prepareStatement()`
- `SQLiteStatement` - Java interface with functions to bind parameters, step through operation and results, and get column data
- `SQLiteGlueConnection` - sample implementation of `SQLiteConnection` that uses `SQLiteGlue` to provide sqlite3 database functionality

When I tried porting the Android version of my Cordova sqlite plugin from sqlite4java to using `SQLiteGlue` and `SQLiteGlueConnection`, I was able to get the basic string manipulation test working in about 45 minutes. It took me a couple hours or so to add a few more functions to get the number of rows affected, fix data bindings, etc. The version at https://github.com/brodybits/Cordova-sqlite-storage-dev/tree/sqlg-test-1 now passes the majority of its tests using my new library classes.

My new library classes are very low level and intended to support higher-level Android database libraries. They also need to be tested further and also documented, maybe in a blog post for now. My idea is to try porting libraries such as sqlite4java, the Android sqlite/sqlcipher library, and/or the database classes from the Android project to use `SQLiteGlue` and `SQLiteGlueConnection` and see how they do against their own test suites.

About Gluegen: I found it a couple years ago from Wikipedia, website at: http://jogamp.org/gluegen/www/. Looks like a similar idea to SWIG but much simpler and a more permissive license on the generator itself.

**Comment by brodybits** [comment]
I expanded the new library classes in https://github.com/sqlc/SQLiteConnector to make two levels of abstraction over https://github.com/sqlg/SQLiteGlue:
- simple interface with accessor functions that return integer status codes
- SQLiteConnection/SQLiteStatement interfaces and classes that add more error checking and throw java.sql.SQLException instances upon receiving a sqlite error result or internally detected error

I was able to embed this in my Cordova sqlite plugin in about a half hour with most of my tests passing here: https://github.com/brodybits/Cordova-sqlite-storage-dev/tree/sqlc-test-1

I was able to get almost all of the Android SQLite-specific database classes to build in an isolated project, but using almost all built-in android.database classes such as `android.database.Cursor`, `android.database.CursorWindow`, etc. from the Android libraries instead of rebuilding these classes. The good news is that all of the native code, which I would like to replace with my wrapper classes, is contained within the `SQLiteConnection` class.

It will take some work to adapt my lower-level wrapper class(es) to fit with the Android SQLite classes, mostly due to their reuse of sqlite statement handles. I am also planning to add another special wrapper class that can help replace the native `SQLiteConnection` functions more easily.

**Comment by brodybits** [comment]
I have used an experimental version of my SQLiteGlue project to build a lighter-weight sqlcipher library for Android and got it working with my Cordova plugin. The `armeabi` shared library takes 469 K (480384 bytes), compared to my build of `libs/armeabi/libsqlcipher_android.so` which takes 2 MB not including `libdatabase_sqlcipher.so` (205 K) and `libstlport_shared.so` (373 K). My Cordova plugin already passes most of its tests when running with my lighter-weight sqlcipher library. Right now my work is committed in the following locations:
- https://github.com/brodybits/sqlg-cipher-core-test-1 - adaptation of SQLiteGlue to build sqlcipher (along with part of libTomCrypt) with the JNI wrapper class (only `sqlite3_key()` accessor for now, still missing `sqlite3_rekey()` accessor)
- https://github.com/brodybits/sqlc-cipher-test-1 - adaptation of SQLiteConnector library to support the sqlcipher-specific additions (only `sqlite3_key()` accessor for now, still missing `sqlite3_rekey()` accessor)
- https://github.com/brodybits/Cordova-sqlcipher-adapter-dev/tree/sqlc-cipher-test-1 - version of my Cordova sqlcipher plugin that uses the new SQLiteGlue/SQLiteConnector classes and library build instead of the "SQLCipher for Android" distribution

Some things that are still missing:
- investigate and attempt to fix remaining test failures in my plugin
- cleanup and documentation of my SQLiteConnector library
- build native library for all architectures, and test 64-bit version(s)
- support for BLOB data (and test it somehow)
- add and test support for rekey

Nice to have:
- ICU-Unicode string manipulation support (limited demand, should be optional if ever supported)

I will continue working on the lighter-weight build of sqlcipher for Android over the coming weeks.

**Comment by developernotes** [comment]
That's great progress @brodybits - thanks for sharing!

**Comment by brodybits** [comment]
I have now published and integrated the following components with my Cordova sqlite plugin:
- [Android-sqlite-native-driver](https://github.com/liteglue/Android-sqlite-native-driver) - builds a Java jar file with native sqlite access libraries, which is invoked through a single `SQLiteNative` Java class that is generated using Gluegen. Interchangeable with jar file generated by [Android-sqlcipher-native-driver](https://github.com/liteglue/Android-sqlcipher-native-driver), to use sqlcipher instead of sqlite
- [Android-sqlite-connector](https://github.com/liteglue/Android-sqlite-connector) - lightweight, pure-Java class library, delivered as a jar file, to access sqlite or sqlcipher database using [Android-sqlite-native-driver](https://github.com/liteglue/Android-sqlite-native-driver) or [Android-sqlcipher-native-driver](https://github.com/liteglue/Android-sqlcipher-native-driver)

[Android-sqlite-native-driver](https://github.com/liteglue/Android-sqlite-native-driver) builds for arm (32-bit), x86 (32-bit), x86_64, and arm-64 (MIPS also possible if anyone dares), and [Android-sqlcipher-native-driver](https://github.com/liteglue/Android-sqlcipher-native-driver) currently buils for arm (32-bit), x86 (32-bit), and x86_64 (arm-64 and MIPS should also be possible).

I have also tested the integration with the sqlcipher version of my plugin, and am planning to publish it once I deal with Circle CI and Travis CI.

I hope this can help others in the future.

**Comment by developernotes** [comment]
Hi @brodybits 

That is great progress, with a nice separation between vanilla SQLite and SQLCipher.  Thanks!

**Comment by brodybits** [comment]
I also started working an a test project to build a newer version of the Android sqlite database source, with a purpose to integrate the Android cursor query interface with my sqlite plugin someday. The test project is in two parts:
- https://github.com/brodybits/Android-sqlcipher-cursor-query-native-driver-test - build a newer version of the CursorWindow code, along with my SQLiteGlue interface generated by Gluegen, to provide native sqlite/sqlcipher access through static Java methods. The native CursorWindow code was updated to match the Android 5.0 source and uses 64-bit integer to store native C/C++ pointers.
- https://github.com/brodybits/Android-sqlite-cursor-query-test - classes from android.database.sqlite, based on the android-m-preview-1 tag, rebuilt to use the native interface provided by Android-sqlcipher-cursor-query-native-driver-test
